internal/irc/handler.go

```diff
@@ -253,41 +256,113 @@ func (h *Handler) Restart() error {
 
-	time.Sleep(4 * time.Second)
+	h.setConnectionStatus()
+
+	if h.haveDisconnected {
+		h.notificationService.Send(domain.NotificationEventIRCReconnected, domain.NotificationPayload{
+			Subject: "IRC Reconnected",
+			Message: fmt.Sprintf("Network: %v", h.network.Name),
+		})
+	}
+	// reset haveDisconnected
+	h.haveDisconnected = true
+
+	h.log.Debug().Msgf("connected to: %v", h.network.Name)
+
+	time.Sleep(2 * time.Second)
 
 	if h.network.NickServ.Password != "" {
-		err := h.HandleNickServIdentify(h.network.NickServ.Password)
-		if err != nil {
+		if err := h.NickServIdentify(h.network.NickServ.Password); err != nil {
 			h.log.Error().Stack().Err(err).Msg("error nickserv")
 			return
 		}
-		identified = true
-	}
 
-	time.Sleep(4 * time.Second)
+		// return and wait for NOTICE of nickserv auth
+		return
+	}
 
-	if h.network.InviteCommand != "" {
-		err := h.handleConnectCommands(h.network.InviteCommand)
-		if err != nil {
+	if h.network.InviteCommand != "" && h.network.NickServ.Password == "" {
+		if err := h.sendConnectCommands(h.network.InviteCommand); err != nil {
 			h.log.Error().Stack().Err(err).Msgf("error sending connect command %v", h.network.InviteCommand)
 			return
 		}
 
-		time.Sleep(1 * time.Second)
+		return
 	}
 
-	if !identified {
-		for _, channel := range h.network.Channels {
-			err := h.HandleJoinChannel(channel.Name, channel.Password)
-			if err != nil {
-				h.log.Error().Stack().Err(err).Msgf("error joining channels %v", err)
+	// join channels if no password or no invite command
+	h.JoinChannels()
+
+}
+
+func (h *Handler) onDisconnect(m ircmsg.Message) {
+	h.log.Debug().Msgf("DISCONNECT")
+
+	h.haveDisconnected = true
+
+	h.resetConnectionStatus()
+
+	h.notificationService.Send(domain.NotificationEventIRCDisconnected, domain.NotificationPayload{
+		Subject: "IRC Disconnected unexpectedly",
+		Message: fmt.Sprintf("Network: %v", h.network.Name),
+	})
+}
+
+func (h *Handler) onNotice(msg ircmsg.Message) {
+	h.log.Debug().Msgf("NOTICE: %v", msg.Nick())
+
+	if msg.Nick() == "NickServ" {
+		h.log.Debug().Msgf("NOTICE from nickserv: %v", msg.Params)
+
+		// params: [test-bot You're now logged in as test-bot]
+		if contains(msg.Params[1], "you're now logged in as") {
+			h.log.Debug().Msgf("NOTICE nickserv logged in: %v", msg.Params)
+
+			// if no invite command, join
+			if h.network.InviteCommand == "" {
+				h.JoinChannels()
 				return
 			}
+
+			// else send connect commands
+			if err := h.sendConnectCommands(h.network.InviteCommand); err != nil {
+				h.log.Error().Stack().Err(err).Msgf("error sending connect command %v", h.network.InviteCommand)
+				return
+			}
+		}
+
+		//[test-bot Invalid parameters. For usage, do /msg NickServ HELP IDENTIFY]
+		if contains(msg.Params[1], "invalid", "help") {
+			h.log.Debug().Msgf("NOTICE nickserv invalid: %v", msg.Params)
+
+			h.client.Send("PRIVMSG", "NickServ", fmt.Sprintf("IDENTIFY %v %v", h.network.NickServ.Account, h.network.NickServ.Password))
```

Comment:
I'm under the impression that this is not universal functionality. It's definitely silly but this should likely old style if the nicks match, and if they don't then pass the nick.

Reply:
This is not totally set in stone yet as it still needs more testing, but this is a backup for if the regular nickserv with just password fails. On Ergo it seem to need both username and password, while I've seen the opposite on some networks where it's not supported to identify with both nick and pass. Fun inconsistencies.

This is a bit different but in [ZNC there's this ](https://github.com/znc/znc/blob/znc-1.8.2/modules/nickserv.cpp) which was referenced by the go-irc maintainer.